web/pdf_outline_viewer.js

```diff
@@ -31,194 +31,187 @@ var DEFAULT_TITLE = '\u2013';
-        return;
-      }
+    if (styleStr) {
+      element.setAttribute('style', styleStr);
+    }
+  }
 
-      var fragment = document.createDocumentFragment();
-      var queue = [{ parent: fragment, items: this.outline }];
-      var hasAnyNesting = false;
-      while (queue.length > 0) {
-        var levelData = queue.shift();
-        for (var i = 0, len = levelData.items.length; i < len; i++) {
-          var item = levelData.items[i];
-
-          var div = document.createElement('div');
-          div.className = 'outlineItem';
-
-          var element = document.createElement('a');
-          this._bindLink(element, item);
-          this._setStyles(element, item);
-          element.textContent =
-            removeNullCharacters(item.title) || DEFAULT_TITLE;
-
-          div.appendChild(element);
-
-          if (item.items.length > 0) {
-            hasAnyNesting = true;
-            this._addToggleButton(div);
-
-            var itemsDiv = document.createElement('div');
-            itemsDiv.className = 'outlineItems';
-            div.appendChild(itemsDiv);
-            queue.push({ parent: itemsDiv, items: item.items });
-          }
-
-          levelData.parent.appendChild(div);
-          outlineCount++;
-        }
-      }
-      if (hasAnyNesting) {
-        this.container.classList.add('outlineWithDeepNesting');
+  /**
+   * Prepend a button before an outline item which allows the user to toggle
+   * the visibility of all outline items at that level.
+   *
+   * @private
+   */
+  _addToggleButton(div) {
+    var toggler = document.createElement('div');
+    toggler.className = 'outlineItemToggler';
+    toggler.onclick = (e) => {
+      e.stopPropagation();
+      toggler.classList.toggle('outlineItemsHidden');
+
+      if (e.shiftKey) {
+        var shouldShowAll = !toggler.classList.contains('outlineItemsHidden');
+        this._toggleOutlineItem(div, shouldShowAll);
       }
+    };
+    div.insertBefore(toggler, div.firstChild);
+  }
 
-      this.container.appendChild(fragment);
+  /**
+   * Toggle the visibility of the subtree of an outline item.
+   *
+   * @param {Element} root - the root of the outline (sub)tree.
+   * @param {boolean} show - whether to show the outline (sub)tree. If false,
+   *   the outline subtree rooted at |root| will be collapsed.
+   *
+   * @private
+   */
+  _toggleOutlineItem(root, show) {
+    this.lastToggleIsShow = show;
+    var togglers = root.querySelectorAll('.outlineItemToggler');
+    for (var i = 0, ii = togglers.length; i < ii; ++i) {
+      togglers[i].classList[show ? 'remove' : 'add']('outlineItemsHidden');
+    }
+  }
 
+  /**
+   * Collapse or expand all subtrees of the outline.
+   */
+  toggleOutlineTree() {
+    if (!this.outline) {
+      return;
+    }
+    this._toggleOutlineItem(this.container, !this.lastToggleIsShow);
+  }
+
+  /**
+   * @param {PDFOutlineViewerRenderParameters} params
+   */
+  render(params = Object.create(null)) {
```

Comment:
Since `params` should never be undefined when the method is called, I think that just `render(params = {}) {` would suffice and it's also slightly more readable in my opinion.